### DISCUSSION TITLE
Air Native Extensions dispatchStatusEventAsync and FREObject

### DISCUSSION THREAD

**Discussion by pol2095** [airsdk/Adobe-Runtime-Support#2563] [Q&A] [answered]
Hello,

Air Native Extensions can dispatch a event asynchronous :
- on Android : [dispatchStatusEventAsync( String code, String level )](https://help.adobe.com/en_US/air/extensions/WS39e706a46ad531be-fd70de2132a8f3874e-8000.html)
- in C : [FREDispatchStatusEventAsync()](https://help.adobe.com/en_US/air/extensions/WSb464b1207c184b14-62b8e11f12937b86be4-7ff5.html)

I suggest to add a new parameter if possible :
`dispatchStatusEventAsync( String code, String level, FREObject object = null )`

Now, I need to call a native method the get the FREObject.

Thanks

**Comment by ajwfrost** [comment, marked as answer]
Hi

Can I check, what you are thinking that this FREObject would be? Is this an arbitrary object that you're trying to pass in to the event handler?

My only concern with this is that the "DispatchStatusEventAsync" code is the one thing that can be called from a separate thread, i.e. to queue an event to be handled within the runtime's main thread. But, you can't create an ActionScript object (aka FREObject) in a background thread, which is why it's just using the native 'string' type values and then passing them around in Java/C until the event dispatcher picks up the request in the runtime thread. And you can't hang on to FREObject types between different calls to ANE functions (as you have no way of knowing whether they'd be garbage collected).

So I think this may only be useful if you wanted to dispatch an event directly from an ANE's normal function i.e. within the execution stack of an `ExtensionContext.call()` method: within that call function, you can create AS3 objects and reference them via FREObject handles, and dispatch these into listeners etc etc. So would this be of interest? Ultimately this could just be a normal event handler rather than a StatusEvent, so we could make something like that a bit easier by providing a new method such as `FREDispatchEvent(string eventType, freobject objectToSend)` (although a more 'normal' case would be to create an appropriate AS3 Event class and to dispatch that, i.e. you could have separate code to create an object that derives from Event and has whatever properties you're after, and then it would just be `FREDispatchEvent(freobject event)` based on the `Event.type` value...)

Let me know what you think!

thanks

**Reply by pol2095** [reply]
I noticed the problem of the thread in the background using FREObject, I give up on this idea.
I continue to use dispatchStatusEventAsync() and after I call a native method to get a FREObject.
Thanks.
